### PR TITLE
feat refactors the chat flow so it has open and close functions

### PR DESCRIPTION
**Pull request by tmoldovan8x8**

### CODE REVIEW

File: modules/API/API.js [hunk 1]
```diff
@@ -15,6 +15,7 @@ import {
 import { parseJWTFromURLParams } from '../../react/features/base/jwt';
 import JitsiMeetJS, { JitsiRecordingConstants } from '../../react/features/base/lib-jitsi-meet';
 import { pinParticipant, getParticipantById } from '../../react/features/base/participants';
+import { openChat } from '../../react/features/chat/';
```
Comment (saghul) [inline modules/API/API.js:18]: can you import this from /actions?

File: react/features/chat/actions.any.js [hunk 2]
```diff
@@ -83,3 +98,14 @@ export function setPrivateMessageRecipient(participant: Object) {
...
     };
 }
+
+/**
+ * Toggles display of the chat panel.
+ *
+ * @returns {Function}
+ */
+export function toggleChat() {
+    return function(dispatch: (Object) => Object) {
```
Comment (saghul) [inline react/features/chat/actions.any.js:108]: Why did you make this action async?

File: react/features/chat/components/AbstractChat.js [hunk 3]
```diff
@@ -92,7 +92,9 @@ export function _mapDispatchToProps(dispatch: Dispatch<any>) {
          */
         _onSendMessage(text: string) {
             dispatch(sendMessage(text));
-        }
+        },
+
+        dispatch
```
Comment (saghul) [inline react/features/chat/components/AbstractChat.js:97]: why is this necessary? When you connect() you shoudl already have dispatch(), otherwise you can just compose a new mapDispatchToProps

File: react/features/chat/components/PrivateMessageButton.js [hunk 4]
```diff
@@ -1,11 +1,11 @@
 // @flow
 
+import { openChat } from '../';
```
Comment (saghul) [inline react/features/chat/components/PrivateMessageButton.js:3]: /actions

File: react/features/chat/components/native/ChatButton.js [hunk 5]
```diff
@@ -55,13 +36,7 @@ class ChatButton extends AbstractButton<Props, *> {
      * @returns {void}
      */
     _handleClick() {
-        if (this.props._showNamePrompt) {
-            this.props._displayNameInputDialog(() => {
```
Comment (saghul) [inline react/features/chat/components/native/ChatButton.js:59]: how is the name requirement handled now?

File: react/features/chat/middleware.js [hunk 6]
```diff
@@ -33,6 +35,8 @@ import {
 import { getUnreadCount } from './functions';
 import { INCOMING_MSG_SOUND_FILE } from './sounds';
 
+import { openChat } from './';
```
Comment (saghul) [inline react/features/chat/middleware.js:38]: /actions

File: react/features/chat/middleware.js [hunk 7]
```diff
@@ -63,14 +68,17 @@ MiddlewareRegistry.register(store => next => action => {
...
         break;
+
     case TOGGLE_CHAT:
-        unreadCount = 0;
-        isOpen = !getState()['features/chat'].isOpen;
+        isOpen = getState()['features/chat'].isOpen;
 
-        if (typeof APP !== 'undefined') {
-            APP.API.notifyChatUpdated(unreadCount, isOpen);
+        if (isOpen) {
```
Comment (saghul) [inline react/features/chat/middleware.js:75]: After your changes, is TOGGLE_CHAT used anywhere?

**Review by saghul — CHANGES_REQUESTED** [review]
Left some comments, nothing major though, great work!

File: react/features/chat/components/native/ChatButton.js (hunk 5 above)
Comment (tmoldovan8x8) [reply react/features/chat/components/native/ChatButton.js:59]: It is handled in the openChat action with, logic similar to what-s here

File: react/features/chat/middleware.js (hunk 7 above)
Comment (tmoldovan8x8) [reply react/features/chat/middleware.js:75]: It is still used by the web that actually has some "toggle functionality", pressing the chat button multiple times has this toggle effect.

Comment (saghul) [reply react/features/chat/middleware.js:75]: 👍

File: react/features/chat/components/web/Chat.js [hunk 8]
```diff
@@ -52,6 +52,9 @@ class Chat extends AbstractChat<Props> {
 
         // Bind event handlers so they are only bound once for every instance.
         this._onChatInputResize = this._onChatInputResize.bind(this);
+
+        // Bind event handlers so they are only bound once for every instance.
```
Comment (saghul) [inline react/features/chat/components/web/Chat.js:56]: No need to duplicate this comment, just put the new line right after the other

File: react/features/chat/middleware.js (hunk 7 above)
Comment (saghul) [inline react/features/chat/middleware.js:75]: Can we move this logic to the action instead? When you do an async action you can check right there, instead of having to go through the middleware. Thsi way we'd get rid of the TOGGLE_CHAT action type.

**Review by saghul — CHANGES_REQUESTED** [review]
A couple of minor comments, excellent work otherwise!

**Review by saghul — APPROVED** [review]
🚀
